### PR TITLE
feat: enhance apigee x and hybrid reporting

### PR DESCRIPTION
**Pull request by g-devpat**
## Description
What's changed, or what was fixed?

Added capability to report the following Apigee X/Hybrid resources:
- [x] Target Servers
- [x] Keystores
- [x] API Products
- [x] Developer Apps
- [x] Developers
- [x] KVM

## Housekeeping
(please check all that apply [x], do not edit the text)
- [ ] I have run all the tests locally and they all pass.
- [ ] I have followed the relevant style guide for my changes.
- [ ] PR requires full pipeline run (Run for changes only by default).

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

**Comment by danistrebel** [comment]
this is great @g-devpat !

**Pro tip:** If you create at least one commit with the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format that goes something like this: `feat: adding additional information to sackmesser report for X and hybrid` it will show up in the release notes and create a minor release that users can refer to.

File: tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+urlencode() {
+    echo "\"${*:1}\"" | jq -r '@uri'
+}
+
+echo "<h3>API Products</h3>" >> "$report_html"
+
+mkdir -p "$export_folder/$organization/config/resources/edge/env/$environment/api-products"
+
+sackmesser list "organizations/$organization/apiproducts"| jq -r -c '.[]|.' | while read -r apiProductName; do
```
Comment (g-greatdevaks) [inline tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh:25]: @g-devpat 
Task: Check requirement (like API Product having white spaces) of urlencode in OPDK code as well.

File: tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    done
+
+if ls "$export_folder/$organization/config/resources/edge/env/$environment/api-products"/*.json 1> /dev/null 2>&1; then
+    jq -n '[inputs]' "$export_folder/$organization/config/resources/edge/env/$environment/api-products"/*.json > "$export_folder/$organization/config/resources/edge/env/$environment/api-products".json
+fi
+
+echo "<div><table id=\"ts-lint\" data-toggle=\"table\" class=\"table\">" >> "$report_html"
+echo "<thead class=\"thead-dark\"><tr>" >> "$report_html"
+echo "<th data-sortable=\"true\" data-field=\"id\">Product Name</th>" >> "$report_html"
+echo "<th data-sortable=\"true\" data-field=\"envs\">Environments</th>" >> "$report_html"
```
Comment (g-greatdevaks) [inline tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh:37]: @g-devpat 
"Proxies" field should be added to this report.

File: tools/apigee-sackmesser/cmd/report/report.sh [hunk 3]
```diff
@@ -423,6 +423,16 @@ if [ "$opdk" == "T" ]; then
     source $opdk_dir/virtualhosts.sh
 fi
 
+if [ "$apiversion" = "google" ]; then
```
Comment (g-greatdevaks) [inline tools/apigee-sackmesser/cmd/report/report.sh:426]: Let's explore consolidation of conditionals in else if format to maintain code coherence.

File: tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    echo "\"${*:1}\"" | jq -r '@uri'
+}
+
+echo "<h3>API Products</h3>" >> "$report_html"
+
+mkdir -p "$export_folder/$organization/config/resources/edge/env/$environment/api-products"
+
+sackmesser list "organizations/$organization/apiproducts"| jq -r -c '.[]|.' | while read -r apiProductName; do
+        sackmesser list "organizations/$organization/apiproducts/$(urlencode "$apiProductName")" > "$export_folder/$organization/config/resources/edge/env/$environment/api-products/$(urlencode "$apiProductName")".json
+        elem_count=$(jq '.entries? | length' "$export_folder/$organization/config/resources/edge/env/$environment/api-products/$(urlencode "$apiProductName")".json)
```
Comment (danistrebel) [inline tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh]: (for all `elem_count` variables in this PR)

I don't see where the `elem_count` values are used

File: tools/apigee-sackmesser/cmd/report/xhybrid/references.sh [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+echo "<h3>References</h3>" >> "$report_html"
+
+mkdir -p "$export_folder/$organization/config/resources/edge/env/$environment/reference"
+
+sackmesser list "organizations/$organization/environments/$environment/references"| jq -r -c '.[]|.' | while read -r referencename; do
```
Comment (danistrebel) [inline tools/apigee-sackmesser/cmd/report/xhybrid/references.sh]: can we be consistent with `referencename` and `referenceName` (L40 below)

File: tools/apigee-sackmesser/cmd/report/xhybrid/developerapps.sh [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+sackmesser list "organizations/$organization/developers" | jq -r -c '.[]|.' | while read -r email; do
+    loginfo "download developer: $email"
+    mkdir "$export_folder/$organization/config/resources/edge/env/$environment/developerapps/$email"
+    sackmesser list "organizations/$organization/developers/$email/apps" | jq -r -c '.[]|.' | while read -r appId; do
+        loginfo "download developer app: $appId for developer: $email"
+        full_app=$(sackmesser list "organizations/$organization/developers/$email/apps/$(urlencode "$appId")")
+        echo "$full_app" | jq 'del(.credentials)' > "$export_folder/$organization/config/resources/edge/env/$environment/developerapps/$appId".json
+        echo "$full_app" | jq -r -c '.credentials[]' | while read -r credential; do
+            appkey=$(echo "$credential" | jq -r '.consumerKey')
+            # echo "$credential" | jq  --arg APP_NAME "$appId" '. + { name: $APP_NAME } | . + { apiProducts: [.apiProducts[]?.apiproduct] }' > "$export_folder/$organization/config/resources/edge/env/$environment/developerapps/$appId".json
```
Comment (danistrebel) [inline tools/apigee-sackmesser/cmd/report/xhybrid/developerapps.sh]: can we remove unused code?

File: tools/apigee-sackmesser/cmd/report/report.sh (hunk 3 above)
Comment (danistrebel) [reply tools/apigee-sackmesser/cmd/report/report.sh]: I'm with Anmol. If possible can we keep the code split between OPDK and x/hybrid as minimal as possible. 
In the past we tried to [harmonize](https://github.com/apigee/devrel/blob/main/tools/apigee-sackmesser/cmd/list/list.sh#L31-L51) the output format of `sackmesser list` and then use the same logic in report based on the unified data model.

**Review by danistrebel — CHANGES_REQUESTED** [review]

File: tools/apigee-sackmesser/cmd/report/xhybrid/developerapps.sh (hunk 6 above)
Comment (g-devpat) [reply tools/apigee-sackmesser/cmd/report/xhybrid/developerapps.sh]: Done

File: tools/apigee-sackmesser/cmd/report/xhybrid/references.sh (hunk 5 above)
Comment (g-devpat) [reply tools/apigee-sackmesser/cmd/report/xhybrid/references.sh]: Done

File: tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh (hunk 4 above)
Comment (g-devpat) [reply tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh]: Done

Comment (danistrebel) [reply tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh]: @g-devpat Can you do the same for the 8 other `elem_count` variables?

File: tools/apigee-sackmesser/cmd/report/report.sh (hunk 3 above)
Comment (danistrebel) [reply tools/apigee-sackmesser/cmd/report/report.sh]: @g-devpat this was resolved without comment. 
I strongly believe we increase the  maintainability cost if we duplicate logic for X and OPDK.

**Comment by danistrebel** [comment]
Re-Opened two conversations above.

/gcbrun

File: tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh (hunk 4 above)
Comment (g-devpat) [reply tools/apigee-sackmesser/cmd/report/xhybrid/apiproducts.sh]: @danistrebel Yes, I did that as part of the previous commit for all the helper files

File: tools/apigee-sackmesser/cmd/report/report.sh (hunk 3 above)
Comment (g-devpat) [reply tools/apigee-sackmesser/cmd/report/report.sh]: @danistrebel Yes, Agreed. earlier we were using seperate helper files for opdk and x which I eliminated as part of this commit 5100fc5. Now we have a single set of helper files for all Apigee versions

**Pull request merged**